presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java

```diff
@@ -108,87 +147,158 @@ private void addTask(RemoteTask task)
+            TaskStatsTracker splitTracker = new TaskStatsTracker("SplitTracker", taskId, nodeTotalPartitionedSplitCount);
+            TaskStatsTracker memoryUsageTracker = new TaskStatsTracker("MemoryTracker", taskId, nodeTotalMemoryUsageInBytes);
+            AccumulatedTaskStatsTracker cpuUsageTracker = new AccumulatedTaskStatsTracker("CpuTracker", taskId, nodeTotalCpuTimePerMillis, windowSizeInMilis);
+            NodeStatsTracker nodeStatsTracker = new NodeStatsTracker(splitTracker::setValue, memoryUsageTracker::setValue, cpuUsageTracker::setValue);
 
-            // when partitionedSplitCountTracker is garbage collected, run the cleanup method on the tracker
-            // Note: tracker can not have a reference to partitionedSplitCountTracker
-            finalizerService.addFinalizer(partitionedSplitCountTracker, tracker::cleanup);
+            // when nodeStatsTracker is garbage collected, run the cleanup method on the tracker
+            // Note: tracker can not have a reference to nodeStatsTracker
+            finalizerService.addFinalizer(nodeStatsTracker, splitTracker::cleanup);
+            finalizerService.addFinalizer(memoryUsageTracker, memoryUsageTracker::cleanup);
+            finalizerService.addFinalizer(cpuUsageTracker, cpuUsageTracker::cleanup);
 
-            return partitionedSplitCountTracker;
+            return nodeStatsTracker;
         }
 
         @ThreadSafe
-        private class TaskPartitionedSplitCountTracker
+        private class TaskStatsTracker
         {
+            private final String stat;
             private final TaskId taskId;
-            private final AtomicInteger localPartitionedSplitCount = new AtomicInteger();
+            private final AtomicLong totalValue;
+            private final AtomicLong value = new AtomicLong();
 
-            public TaskPartitionedSplitCountTracker(TaskId taskId)
+            public TaskStatsTracker(String stat, TaskId taskId, AtomicLong totalValue)
             {
+                this.stat = requireNonNull(stat, "stat is null");
                 this.taskId = requireNonNull(taskId, "taskId is null");
+                this.totalValue = requireNonNull(totalValue, "totalValue is null");
             }
 
-            public synchronized void setPartitionedSplitCount(int partitionedSplitCount)
+            public synchronized void setValue(long value)
             {
-                if (partitionedSplitCount < 0) {
-                    int oldValue = localPartitionedSplitCount.getAndSet(0);
-                    nodeTotalPartitionedSplitCount.addAndGet(-oldValue);
-                    throw new IllegalArgumentException("partitionedSplitCount is negative");
+                if (value < 0) {
+                    long oldValue = this.value.getAndSet(0L);
+                    totalValue.addAndGet(-oldValue);
+                    throw new IllegalArgumentException(stat + " is negative");
```

Comment:
Can you help me understand the logic here .
If we are throwing an IllegalArg exception wondering why we need to set a negative value and then throw, shouldn't we just throw immediately.?

Reply:
The value is invalid. We need to fix the accumulated (per node) value for this task first and throw and exception.